### ISSUE TITLE
Support leave functionality in simpleTraverse

### ISSUE THREAD

**Issue by gaborantal** [typescript-eslint/typescript-eslint#3647]
Currently, in SimpleTraverseOptions [1], only "enter" functionality is supported. Although the "leave" functionality may often be useful (even necessary) in a traversal.

[1]: https://github.com/typescript-eslint/typescript-eslint/blob/84fff3575282f09683385e855c20df3c261ce313/packages/typescript-estree/src/simple-traverse.ts#L19

- [x] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```TS
23
```

```TS
import * as tsEstree from '@typescript-eslint/typescript-estree';

try {
    let rawAst = tsEstree.parse("23");

    tsEstree.simpleTraverse(rawAst, {
        enter: function (node) {
            console.log("Hello");
        },
        leave: function (node) {
            console.log("Bye");
        }
    })

} catch (e) {
    console.log(e)
}
```

**Expected Result**

Hello
Hello
Hello
Bye
Bye
Bye

**Actual Result**

Hello
Hello
Hello

**Additional Info**

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `4.28.4` |
| `TypeScript`                           | `4.3.5` |
| `node`                                 | `14 LTS` |

**Comment by bradzacher** [comment]
what's your usecase?
Could you explain more about what you're trying to do?

`simpleTraverse` isn't really a public API - it's really only exported from the module so we can consume it internally. Which is why there are no JSDoc comments nor is there any mention of it in the markdown documentation.

**Comment by gaborantal** [comment]
Sure. I'm working on a code analysis tool, and we implemented a variable usage detector. We are using a self-implemented scope manager that handles scoping and helps us to determine variable usages. When we leave a node, we modify the scope manager's state. I'm not totally convinced about our solution handles everything well, but it works most of the time pretty good. However, if you have any other ideas or I missed some package that would help me to do that, please let me know.

**Comment by bradzacher** [comment]
Why not just leverage our scope analyser which does exactly what you want in terms of finding variables and references? https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/scope-manager

And we have additional code for detecting variable usage:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/util/collectUnusedVariables.ts

There's some additional logic located in https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-unused-vars.ts because it relies on complex selectors or configurability. All of the "complex logic" is to do with `namespace`s and `.d.ts` files so it can pretty well be ignored, however.

**Comment by bradzacher** [comment]
As simpleTraverse is intentionally light-weight and simple and is an internal tool for us to reuse internally for operating on the AST - we won't be adding this.

There are better methods for supporting the request - as mentioned.
